sdk/eventhub/azure-eventhub/azure/eventhub/_producer_client.py

```diff
@@ -14,13 +14,45 @@
 from ._producer import EventHubProducer
 from ._constants import ALL_PARTITIONS
 from ._common import EventDataBatch, EventData
+from ._configuration import PartitionPublishingConfiguration
+from ._utils import validate_producer_client_partition_config
 
 if TYPE_CHECKING:
     from azure.core.credentials import TokenCredential
+    PartitionPublishingConfigType = Optional[Union[PartitionPublishingConfiguration, dict]]
 
 _LOGGER = logging.getLogger(__name__)
 
 
+def validate_outgoing_event_data(
+    event_data_batch,
+    partition_id=None,
+    partition_key=None,
+    is_idempotent_publishing=False
+):
+    # type: (Union[EventDataBatch, List[EventData]], Optional[str], Optional[str], bool) -> None
+    # Validate the input of EventHubProducerClient.send_batch
+    # pylint:disable=protected-access
+    if isinstance(event_data_batch, EventDataBatch):
+        if is_idempotent_publishing:
+            if event_data_batch._partition_id is None:
+                raise ValueError("The EventDataBatch object must have the partition_id set when performing "
+                                 "idempotent publishing. Please create an EventDataBatch object with partition_id.")
+            if event_data_batch.starting_published_sequence_number is not None:
+                raise ValueError("EventDataBatch object that has already been published by idempotent producer"
+                                 "cannot be published again. Please create a new object.")
+        if partition_id or partition_key:
+            raise TypeError("partition_id and partition_key should be None when sending an EventDataBatch "
+                            "because type EventDataBatch itself may have partition_id or partition_key")
```

Comment:
Is raising this error a change in behaviour for anyone? Or were we previously raising it somewhere else?

Reply:
the latter one, it is still raised in the `create_batch` method -- I did some refactor to reduce code duplication, extracting the same logic from the `create_batch` from both of sync and async producer client and put them in a helper method `validate_outgoing_event_data` which is called by `create_batch` and could be used by both of the clients.